File: bindings/python/python/fastlowess/_gpu_installer.py
"""One-time downloader/installer for the opt-in GPU-enabled fastlowess build.

The GPU backend (wgpu) is not included in the wheels published to PyPI. This
module fetches a prebuilt GPU-enabled wheel from the matching GitHub Release
and installs it in place of the current (CPU-only) installation.
"""

from __future__ import annotations

import json
import platform
import subprocess
import sys
import tempfile
import urllib.error
import urllib.request
from pathlib import Path

from .__version__ import __version__

_REPO = "thisisamirv/lowess-project"
_API_URL = f"https://api.github.com/repos/{_REPO}/releases/tags/v{__version__}"
_USER_AGENT = f"fastlowess/{__version__} (+https://github.com/{_REPO})"


def gpu_available() -> bool:
    """Return True if this installation was built with the GPU backend enabled."""
    from . import _core

    return _core.gpu_enabled()


def _current_platform_tag() -> str:
    if sys.platform.startswith("win"):
        return "windows"
    if sys.platform == "darwin":
        return "macos"
    return "linux"


def _current_arch_tag() -> str:
    machine = platform.machine().lower()
    if machine in ("amd64", "x86_64"):
        return "x86_64"
    if machine in ("arm64", "aarch64"):
        return "aarch64"
    return machine


def _asset_matches_platform(name: str) -> bool:
    name = name.lower()
    plat = _current_platform_tag()
    arch = _current_arch_tag()

    if plat == "windows":
        if arch == "aarch64":
            return "win_arm64" in name
        return "win_amd64" in name or "win32" in name
    if plat == "macos":
        if "macosx" not in name:
            return False
        if arch == "aarch64":
            return "arm64" in name or "universal2" in name
        return "x86_64" in name or "universal2" in name
    # linux
    if "linux" not in name:
        return False
    return arch in name


def _asset_matches_python(name: str) -> bool:
    name = name.lower()
    if "abi3" in name:
        return True
    tag = f"cp{sys.version_info.major}{sys.version_info.minor}"
    return tag in name


def _fetch_release_assets() -> list[dict]:
    req = urllib.request.Request(
        _API_URL,
        headers={"User-Agent": _USER_AGENT, "Accept": "application/vnd.github+json"},
    )
    try:
        with urllib.request.urlopen(req, timeout=30) as resp:
            data = json.load(resp)
    except urllib.error.HTTPError as e:
        raise RuntimeError(
            f"Could not find a GitHub release for fastlowess v{__version__} "
            f"({_API_URL}): {e}"
        ) from e
    return data.get("assets", [])


def _find_gpu_wheel_asset() -> dict:
    assets = _fetch_release_assets()
    candidates = [
        a
        for a in assets
        if a["name"].endswith(".whl")
        and "gpu" in a["name"].lower()
        and _asset_matches_python(a["name"])
        and _asset_matches_platform(a["name"])
    ]
    if not candidates:
        raise RuntimeError(
            "No matching GPU wheel found in the fastlowess "
            f"v{__version__} release for this platform/Python version. "
            "You may need to build it locally instead — see "
            "https://lowess.readthedocs.io/api/python/#gpu-acceleration"
        )
    return candidates[0]


def install_gpu(yes: bool = False) -> None:
    """Download and install the GPU-enabled fastlowess build for this platform.

    Fetches a prebuilt wheel (built with the ``gpu`` Cargo feature) from the
    matching GitHub Release over HTTPS and installs it in place of the
    current (CPU-only) installation via pip. Restart the Python process
    afterwards — a loaded native extension cannot be swapped in place.

    Parameters
    ----------
    yes : bool
        Skip the interactive confirmation prompt. Must be True when stdin
        is not an interactive terminal.
    """
    if gpu_available():
        print("GPU backend is already installed.")
        return

    asset = _find_gpu_wheel_asset()
    size_mb = asset.get("size", 0) / (1024 * 1024)

    if not yes:
        if not sys.stdin.isatty():
            raise RuntimeError(
                "install_gpu() requires confirmation. Pass yes=True to "
                "proceed non-interactively."
            )
        answer = input(
            f"Download and install {asset['name']} ({size_mb:.1f} MB) from "
            f"github.com/{_REPO}? [y/N] "
        )
        if answer.strip().lower() not in ("y", "yes"):
            print("Aborted.")
            return

    url = asset["browser_download_url"]
    print(f"Downloading {url} ...")
    req = urllib.request.Request(url, headers={"User-Agent": _USER_AGENT})
    with tempfile.TemporaryDirectory() as tmp:
        wheel_path = Path(tmp) / asset["name"]
        with (
            urllib.request.urlopen(req, timeout=300) as resp,
            open(wheel_path, "wb") as f,
        ):
            f.write(resp.read())

        print("Installing...")
        subprocess.run(
            [
                sys.executable,
                "-m",
                "pip",
                "install",
                "--force-reinstall",
                "--no-deps",
                str(wheel_path),
            ],
            check=True,
        )

    print(
        "GPU backend installed. Restart your Python process/kernel for the "
        "change to take effect."
    )


def _cli() -> None:
    """Entry point for the `fastlowess-install-gpu` console script."""
    install_gpu()

File: bindings/go/fastlowess/gpu_installer.go
package fastlowess

// One-time downloader for the opt-in GPU-enabled fastlowess_go static
// library. Unlike Python/Node.js/R/C++/Julia, cgo links this library
// statically at build time, so there's no runtime install: this only
// downloads the library, then instructs the caller to rebuild with
// CGO_LDFLAGS pointing at it.

import (
	"bufio"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"time"
)

const gpuRepo = "thisisamirv/lowess-project"

// GPU artifacts across all versions live in this one perpetual release
// instead of cluttering each version's own release page; the source version
// is embedded in each asset's filename instead.
const gpuReleaseTag = "gpu-builds"

// platformArchTag returns the "<platform>-<arch>" suffix release-gpu.yml
// uses for this binding's GPU static library, or "" if this platform/arch
// combination isn't built.
func platformArchTag() string {
	switch runtime.GOOS {
	case "linux":
		switch runtime.GOARCH {
		case "amd64":
			return "linux-x86_64"
		case "arm64":
			return "linux-arm64"
		}
	case "darwin":
		switch runtime.GOARCH {
		case "amd64":
			return "macos-x86_64"
		case "arm64":
			return "macos-aarch64"
		}
	case "windows":
		switch runtime.GOARCH {
		case "amd64":
			return "windows-x86_64"
		case "arm64":
			return "windows-arm64"
		}
	}
	return ""
}

// InstallGPU installs a GPU-enabled fastlowess_go static library for this
// platform, saving it to ~/.fastlowess/gpu/libfastlowess_go.a.
//
// Pass yes=true to skip the interactive y/N confirmation prompt; this is
// required when stdin is not an interactive terminal.
//
// Pass a non-empty localPath to install a GPU-enabled library already built
// locally (e.g. via `cargo build -p fastlowess-go --profile release-c
// --features gpu`) instead of downloading one from the matching GitHub
// Release — useful for testing the installer itself, or installing an
// unreleased build.
func InstallGPU(yes bool, localPath string) error {
	if GPUEnabled() {
		fmt.Println("GPU backend is already active.")
		return nil
	}

	home, err := os.UserHomeDir()
	if err != nil {
		home = "."
	}
	dir := filepath.Join(home, ".fastlowess", "gpu")
	dest := filepath.Join(dir, "libfastlowess_go.a")

	if localPath != "" {
		if _, err := os.Stat(localPath); err != nil {
			return fmt.Errorf("no such file: %s", localPath)
		}
		if !yes {
			fmt.Printf("Install %s in place of the current build? [y/N] ", localPath)
			reader := bufio.NewReader(os.Stdin)
			answer, _ := reader.ReadString('\n')
			answer = strings.TrimSpace(strings.ToLower(answer))
			if answer != "y" && answer != "yes" {
				fmt.Println("Aborted.")
				return nil
			}
		}

		if err := os.MkdirAll(dir, 0o755); err != nil {
			return fmt.Errorf("failed to create %s: %w", dir, err)
		}
		fmt.Printf("Installing %s ...\n", localPath)
		if err := copyGPULibrary(localPath, dest); err != nil {
			return fmt.Errorf("failed to copy %s: %w", localPath, err)
		}
		printRebuildInstructions(dest, dir)
		return nil
	}

	tag := platformArchTag()
	if tag == "" {
		return fmt.Errorf("no prebuilt GPU library available for %s/%s; build from source instead: "+
			"cargo build -p fastlowess-go --profile release-c --features gpu", runtime.GOOS, runtime.GOARCH)
	}

	assetName := fmt.Sprintf("libfastlowess_go-gpu-v%s-%s.a", version, tag)
	url := fmt.Sprintf("https://github.com/%s/releases/download/%s/%s", gpuRepo, gpuReleaseTag, assetName)

	if !yes {
		fmt.Printf("Download and install %s from github.com/%s? [y/N] ", assetName, gpuRepo)
		reader := bufio.NewReader(os.Stdin)
		answer, _ := reader.ReadString('\n')
		answer = strings.TrimSpace(strings.ToLower(answer))
		if answer != "y" && answer != "yes" {
			fmt.Println("Aborted.")
			return nil
		}
	}

	if err := os.MkdirAll(dir, 0o755); err != nil {
		return fmt.Errorf("failed to create %s: %w", dir, err)
	}

	fmt.Printf("Downloading %s ...\n", url)
	if err := downloadGPULibrary(url, dest); err != nil {
		return fmt.Errorf("failed to download %s: %w\n"+
			"a matching GPU build may not exist for this platform/version yet", url, err)
	}

	printRebuildInstructions(dest, dir)
	return nil
}

func printRebuildInstructions(dest, dir string) {
	fmt.Printf("GPU library installed at %s.\n", dest)
	fmt.Println("A running program can't swap a statically-linked library, so rebuild with " +
		"CGO_LDFLAGS pointing at it, e.g.:")
	switch runtime.GOOS {
	case "windows":
		fmt.Printf("  set CGO_LDFLAGS=-L%s -lfastlowess_go -lws2_32 -luserenv -lbcrypt -lntdll -lpthread\n", dir)
	case "darwin":
		fmt.Printf("  export CGO_LDFLAGS=\"-L%s -lfastlowess_go\"\n", dir)
	default:
		fmt.Printf("  export CGO_LDFLAGS=\"-L%s -lfastlowess_go -lm -ldl -lpthread\"\n", dir)
	}
	fmt.Println("then `go build ./...`.")
}

func copyGPULibrary(src, dest string) error {
	in, err := os.Open(src)
	if err != nil {
		return err
	}
	defer func() { _ = in.Close() }()

	tmp := dest + ".tmp"
	out, err := os.Create(tmp)
	if err != nil {
		return err
	}
	if _, err := io.Copy(out, in); err != nil {
		_ = out.Close()
		_ = os.Remove(tmp)
		return err
	}
	if err := out.Close(); err != nil {
		_ = os.Remove(tmp)
		return err
	}
	return os.Rename(tmp, dest)
}

func downloadGPULibrary(url, dest string) error {
	client := &http.Client{Timeout: 5 * time.Minute}
	resp, err := client.Get(url)
	if err != nil {
		return err
	}
	defer func() { _ = resp.Body.Close() }()
	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("HTTP %d", resp.StatusCode)
	}

	tmp := dest + ".tmp"
	f, err := os.Create(tmp)
	if err != nil {
		return err
	}
	if _, err := io.Copy(f, resp.Body); err != nil {
		_ = f.Close()
		_ = os.Remove(tmp)
		return err
	}
	if err := f.Close(); err != nil {
		_ = os.Remove(tmp)
		return err
	}
	return os.Rename(tmp, dest)
}
